compiler/rustc_trait_selection/src/opaque_types.rs

```diff
@@ -1095,18 +1100,40 @@ impl<'a, 'tcx> Instantiator<'a, 'tcx> {
         let tcx = infcx.tcx;
 
         // Use the same type variable if the exact same opaque type appears more
-        // than once in the return type (e.g., if it's passed to a type alias).
-        if let Some(opaque_defn) = self.opaque_types.get(&def_id) {
-            debug!("instantiate_opaque_types: returning concrete ty {:?}", opaque_defn.concrete_ty);
-            return opaque_defn.concrete_ty;
+        // than once in a function (e.g., if it's passed to a type alias).
+        if let Some(opaque_defn) = self.opaque_types.get_mut(&def_id) {
```

Comment:
I feel like the *right* behavior here is to key by def-id+substs. We know that the substs are always universally quantified placeholders, right?

Reply:
The substs here are concrete types from the function. I don't understand what universally quantified placeholders mean here

Reply:
Keying by DefId+substs makes sense though. We'll get two type variables, and then we can figure out whether they are compatible in the same pass that does it cross-function

Reply:
"Universally quantified placeholders" is jargon for the "generic parameters on the fn that we are type-checking". i.e., when we treat a parameter `T` as a kind of placeholder for "some type". "universally quantified' is because there is a "forall<T>" quantifier on the function.